src/NHibernate/Async/Loader/Loader.cs

```diff
@@ -0,0 +1,1211 @@
+			IPersistenceContext persistenceContext = session.PersistenceContext;
+
+			object collectionRowKey = await (persister.ReadKeyAsync(rs, descriptor.SuffixedKeyAliases, session, cancellationToken)).ConfigureAwait(false);
+
+			if (collectionRowKey != null)
+			{
+				// we found a collection element in the result set
+
+				if (Log.IsDebugEnabled)
+				{
+					Log.Debug("found row of collection: " + MessageHelper.CollectionInfoString(persister, collectionRowKey));
+				}
+
+				object owner = optionalOwner;
+				if (owner == null)
+				{
+					owner = persistenceContext.GetCollectionOwner(collectionRowKey, persister);
+					if (owner == null)
+					{
+						//TODO: This is assertion is disabled because there is a bug that means the
+						//      original owner of a transient, uninitialized collection is not known 
+						//      if the collection is re-referenced by a different object associated 
+						//      with the current Session
+						//throw new AssertionFailure("bug loading unowned collection");
+					}
+				}
+				IPersistentCollection rowCollection =
+					persistenceContext.LoadContexts.GetCollectionLoadContext(rs).GetLoadingCollection(persister, collectionRowKey);
+
+				if (rowCollection != null)
+				{
+					await (rowCollection.ReadFromAsync(rs, persister, descriptor, owner, cancellationToken)).ConfigureAwait(false);
+				}
+			}
+			else if (optionalKey != null)
+			{
+				// we did not find a collection element in the result set, so we
+				// ensure that a collection is created with the owner's identifier,
+				// since what we have is an empty collection
+
+				if (Log.IsDebugEnabled)
+				{
+					Log.Debug("result set contains (possibly empty) collection: " + MessageHelper.CollectionInfoString(persister, optionalKey));
+				}
+				persistenceContext.LoadContexts.GetCollectionLoadContext(rs).GetLoadingCollection(persister, optionalKey);
+				// handle empty collection
+			}
+
+			// else no collection element, but also no owner
+		}
+
+		/// <summary>
+		/// Read a row of <c>EntityKey</c>s from the <c>DbDataReader</c> into the given array.
+		/// </summary>
+		/// <remarks>
+		/// Warning: this method is side-effecty. If an <c>id</c> is given, don't bother going
+		/// to the <c>DbDataReader</c>
+		/// </remarks>
+		private async Task<EntityKey> GetKeyFromResultSetAsync(int i, IEntityPersister persister, object id, DbDataReader rs, ISessionImplementor session, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			object resultId;
+
+			// if we know there is exactly 1 row, we can skip.
+			// it would be great if we could _always_ skip this;
+			// it is a problem for <key-many-to-one>
+
+			if (IsSingleRowLoader && id != null)
+			{
+				resultId = id;
+			}
+			else
+			{
+				IType idType = persister.IdentifierType;
+				resultId = await (idType.NullSafeGetAsync(rs, EntityAliases[i].SuffixedKeyAliases, session, null, cancellationToken)).ConfigureAwait(false);
```

Comment:
> I do not know if this is actually an issue we need to address, having those "synchronous async calls", or not.

Depends on how often are they called. I've [created](https://dotnetfiddle.net/rIk7jf) a simple performance test so see how much slower an async method with an already completed task is.
If the call is heavily used we could alter it to conditionally call the async counterpart. Example:
```
resultId = (/*our condition*/) ? await (idType.NullSafeGetAsync(...)).ConfigureAwait(false) : idType.NullSafeGet(...);
```
But this can be done only if the condition is known.
Currently the generator is not able to do the above alternation, but I can add support to it if we see that this can significantly improve the performance.